src/pid/environ.rs

```diff
@@ -8,59 +8,148 @@ use std::path::Path;
-        })
+        if self.leftover.len() == 0 {
+            return None;
+        }
+        let (leftover, item) = match self.parse() {
+            None => return None,
+            Some(Err(x)) => return Some(Err(x)),
+            Some(Ok(x)) => x,
+        };
+        self.leftover = leftover;
+        Some(Ok(item))
+    }
+}
+
+impl<'a> IntoIterator for &'a Environ {
+    type Item = Result<(OsString, OsString)>;
+    type IntoIter = EnvironIter<'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        EnvironIter {
+            environ: self,
+            leftover: &self.data,
+        }
     }
 }
 
+/// Extracts name of a variable. Also consumes a delimiter.
+fn get_name(src: &[u8]) -> IResult<&[u8], OsString> {
+    // Calculate position of the equal sign.
+    let pos = match src.iter().skip(1).position(|c| c == &b'=') {
+        Some(p) => p,
+        None => return IResult::Error(error_position!(nom::ErrorKind::Custom(0), src)),
+    };
+    IResult::Done(&src[pos + 2..], from_bytes(&src[..pos + 1]))
+}
+
+#[test]
+fn test_get_name() {
+    let src = &b"FOO=BAR"[..];
+    assert_eq!(get_name(src), IResult::Done(&b"BAR"[..], "FOO".into()));
+    let src = &b"FOO="[..];
+    assert_eq!(get_name(src), IResult::Done(&b""[..], "FOO".into()));
+    let src = &b"=FOO=BAR"[..];
+    assert_eq!(get_name(src), IResult::Done(&b"BAR"[..], "=FOO".into()));
+    let src = &b"=FOO="[..];
+    assert_eq!(get_name(src), IResult::Done(&b""[..], "=FOO".into()));
+}
+
+/// Executes `take_until_and_consume` and maps its result to OsString.
+named_args!(
+    until<'a>(sep: &'a str)<OsString>,
+    map!(take_until_and_consume!(sep), from_bytes)
+);
+
+/// Parses 'key=value" pair.
+named!(
+    pair<&[u8], (OsString, OsString)>,
+    tuple!(get_name, apply!(until, "\0"))
+);
+
+#[test]
+fn test_pair() {
+    let source = &b"FOO=BAR\0123"[..];
+    assert_eq!(
+        pair(source),
+        IResult::Done(&b"123"[..], ("FOO".into(), "BAR".into()))
+    );
+    let source = &b"FOO=\0123"[..];
+    assert_eq!(
+        pair(source),
+        IResult::Done(&b"123"[..], ("FOO".into(), "".into()))
+    );
+    let source = &b"=FOO=BAR\0-"[..];
+    assert_eq!(
+        pair(source),
+        IResult::Done(&b"-"[..], ("=FOO".into(), "BAR".into()))
+    );
+    let source = &b"=FOO=\0-"[..];
+    assert_eq!(
+        pair(source),
+        IResult::Done(&b"-"[..], ("=FOO".into(), "".into()))
+    );
+}
+
+
 #[test]
 fn test_iter() {
-    let mut env = Environ { data: b"key1=val1\0key2=val2\0".to_vec() };
-    assert_eq!(env.next().unwrap().unwrap(), ("key1".into(), "val1".into()));
-    assert_eq!(env.next().unwrap().unwrap(), ("key2".into(), "val2".into()));
-    assert!(env.next().is_none());
+    let env = Environ { data: b"key1=val1\0=key2=val2\0key3=val3\0".to_vec() };
+    let mut iter = env.into_iter();
+
+    let next = iter.next();
+    assert!(next.is_some());
+    let next = next.unwrap();
+    assert!(next.is_ok());
+    let next = next.unwrap();
```

Comment:
At first I was doing this unwraps w/o `is_*` checks and it ended up in a mess when an actual error has happened.